File: src/foam/dao/index/PersistedIndexTest.java
package foam.dao.index;

import foam.nanos.auth.Country;
import org.apache.commons.text.RandomStringGenerator;

import java.io.IOException;

import static org.apache.commons.text.CharacterPredicates.DIGITS;
import static org.apache.commons.text.CharacterPredicates.LETTERS;

public class PersistedIndexTest {

  protected static final int SAMPLE_SIZE = 1000000;
  protected static String[] samples = new String[SAMPLE_SIZE];
  protected static RandomStringGenerator generator =
      new RandomStringGenerator.Builder()
          .withinRange('0', 'z')
          .filteredBy(LETTERS, DIGITS)
          .build();

  public static void main(String[] args) throws IOException {

    // generate random strings of length 50
    for ( int i = 0; i < SAMPLE_SIZE; i++ ) {
      samples[i] = generator.generate(50);
    }

    PersistedIndex index = new PersistedIndex("test", new TreeIndex(Country.CODE));
    long start = System.nanoTime();

    for ( int i = 0; i < SAMPLE_SIZE; i++ ) {
      index.wrap(samples[i]);
    }

    long end = System.nanoTime();
    long duration = (end - start);
    System.out.println("Duration is " + (((double) duration) / 1000000000.0) + " seconds");
  }
}

File: src/foam/dao/index/PersistedIndexTest.js
foam.CLASS({
  package: 'foam.dao.index',
  name: 'PersistedIndexTest',
  extends: 'foam.nanos.test.Test',

  methods: [
    {
      name: 'runTest',
      javaCode: `
        try {
          foam.dao.index.TreeIndex index = new foam.dao.index.TreeIndex(foam.nanos.auth.Country.CODE);
          foam.dao.index.PersistedIndex pindex = new foam.dao.index.PersistedIndex("testindex", index);

          foam.dao.index.PersistedState w1 = (foam.dao.index.PersistedState) pindex.wrap("hello world");
          pindex.flush(w1);
          test(w1.getPosition() == 0, "position is 0");
          test("hello world".equals(pindex.unwrap(w1)), "state is hello world");

          foam.dao.index.PersistedState w2 = (foam.dao.index.PersistedState) pindex.wrap("test123");
          pindex.flush(w2);
          test(w2.getPosition() == 18, "position is 18");
          test("test123".equals(pindex.unwrap(w2)), "state is test123");
        } catch ( Throwable t ) {
          throw new RuntimeException(t);
        }
      `
    }
  ]
});
